apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/TransactionPool.cs

```diff
@@ -0,0 +1,247 @@
+        {
+            return RunFinalMethodAsync(transaction, session,
+                info => info.SpannerClient.CommitAsync(session.SessionName, info.ActiveTransaction.Id, mutations));
+        }
+
+        /// <summary>
+        /// Rolls back a transaction and internally marks the transaction as used.
+        /// </summary>
+        /// <param name="transaction"></param>
+        /// <param name="session"></param>
+        /// <returns></returns>
+        public static Task RollbackAsync(this Transaction transaction, Session session)
+        {
+            return RunFinalMethodAsync(transaction, session,
+                info => info.SpannerClient.RollbackAsync(session.SessionName, info.ActiveTransaction.Id));
+        }
+
+        private static async Task RunFinalMethodAsync(Transaction transaction, Session session, Func<SessionInfo, Task> commitOrRollbackAction)
+        {
+            SessionInfo info;
+            if (s_sessionInfoTable.TryGetValue(session, out info))
+            {
+                //we should have already waited, but extra checking cannot hurt.
+                await info.WaitForPreWarmAsync().ConfigureAwait(false);
+                if (info.ActiveTransaction == null
+                    || info.ActiveTransaction.Id.IsEmpty
+                    || !Equals(info.ActiveTransaction.Id, transaction.Id))
+                {
+                    throw new InvalidOperationException("The transaction being committed was not found to have a valid entry.");
+                }
+
+                try
+                {
+                    await commitOrRollbackAction(info);
+                }
+                finally
+                {
+                    Session ignored;
+                    s_activeTransactionTable.TryRemove(info.ActiveTransaction.Id, out ignored);
+                }
+            }
+            else
+            {
+                throw new ArgumentException(
+                    "The given transaction was not found in the transaction pool. Only "
+                    +
+                    "use this method on Transactions returned from TransactionPoool.CreateTransactionAsync.");
+            }
+        }
+
+        private class SessionInfo
+        {
+            public Transaction ActiveTransaction { get; set; }
+            public TransactionOptions ActiveTransactionOptions { get; set; }
+            public SpannerClient SpannerClient { get; set; }
+            private Task PreWarmTask { get; set; }
+
+            public async Task WaitForPreWarmAsync()
+            {
+                try
+                {
+                    Task prewarmTask = PreWarmTask;
+                    if (prewarmTask != null && !prewarmTask.IsCompleted)
+                    {
+                        await prewarmTask.ConfigureAwait(false);
+                    }
+                    PreWarmTask = null;
+                }
+                catch (Exception e)
```

Comment:
The thing is... this is a noncritical function.  It's ok to fail.

This is functionally equivalent  to the IgnoreExceptions extension method that Toub suggests:
 https://blogs.msdn.microsoft.com/pfxteam/2009/06/01/tasks-and-unhandled-exceptions/

We could use that exception instead of catching.  That would also give us a central place where we can do logging or anything else to capture that the exception happened.